### PR TITLE
Implement the "pretty printing feature", issue #20

### PR DESCRIPTION
**Pull request by lelit**
This implements a `PrettyPrinter` class with a set of node-specific functions that serializes a _syntax tree_ back into a textual representation: it currently handles all basic _DML_ statements. To test the functionality, it compares the syntax tree of the original SQL to that of the reformatted representation.

It also implements several new `Node` types: as requested in #6, I'm willing to rebase all these commits into two separate branches, if you prefer, but I'd like to have your bless before going that route. In that case, given that all new Nodes are already heavily tested by this functionality, I'd like to know if you think I should still write standalone test cases for them.

Thanks in advance for the review!

### CODE REVIEW

**Comment by lelit** [comment]
Uhm, I used `pytest` as recommended in the README, I didn't realize it's not being used by Travis...

**Comment by lelit** [comment]
Is there any particular reason to prefer the standard `unittest` runner over `pytest`?

**Comment by alculquicondor** [comment]
There is no particular reason. Please feel free to add the dependency in travis if needed. Also, since we are going to do separate PRs for the Nodes and then the PP feature, we should have tests for the former. They don't need to be very complex.

File: psqlparse/nodes/value.py [hunk 1]
```diff
@@ -35,6 +35,26 @@ def val(self):
         return self.str
 
 
+class Literal(String):
```
Comment (alculquicondor) [inline psqlparse/nodes/value.py]: This file should mimic https://github.com/postgres/postgres/blob/master/src/include/nodes/value.h
The only allowed ttypes are the following: https://github.com/postgres/postgres/blob/master/src/include/nodes/nodes.h#L278-L286

File: psqlparse/nodes/parsenodes.py [hunk 2]
```diff
@@ -85,7 +93,7 @@ class UpdateStmt(Statement):
 
     def __init__(self, obj):
         self.relation = build_from_item(obj, 'relation')
-        self.target_list = build_from_item(obj, 'targetList')
+        self.target_list = build_from_item(obj, 'targetList', 'ResTargetUpdate')
```
Comment (alculquicondor) [inline psqlparse/nodes/parsenodes.py]: why is this override needed?

File: psqlparse/nodes/parsenodes.py [hunk 3]
```diff
@@ -14,6 +17,11 @@ def __str__(self):
 
 class SelectStmt(Statement):
 
+    OP_NONE = 0
```
Comment (alculquicondor) [inline psqlparse/nodes/parsenodes.py]: please use this constants names: https://github.com/postgres/postgres/blob/d363d42bb9a4399a0207bd3b371c966e22e06bd3/src/include/nodes/parsenodes.h#L1490

File: psqlparse/nodes/parsenodes.py [hunk 4]
```diff
@@ -159,8 +167,8 @@ def tables(self):
 class CommonTableExpr(Node):
 
     def __init__(self, obj):
-        self.ctename = obj.get('ctename')
-        self.aliascolnames = build_from_item(obj, 'aliascolnames')
+        self.ctename = Name.from_string(obj.get('ctename'))
```
Comment (alculquicondor) [inline psqlparse/nodes/parsenodes.py]: This class is going away

File: psqlparse/nodes/parsenodes.py [hunk 5]
```diff
@@ -212,6 +220,10 @@ def tables(self):
         return set()
 
 
+class ResTargetUpdate(ResTarget):
```
Comment (alculquicondor) [inline psqlparse/nodes/parsenodes.py]: please remove this class

File: psqlparse/nodes/parsenodes.py [hunk 6]
```diff
@@ -251,9 +263,26 @@ def tables(self):
 
 class AExpr(Node):
 
+    KIND_OP = 0                 # normal operator
```
Comment (alculquicondor) [inline psqlparse/nodes/parsenodes.py]: constant names here https://github.com/postgres/postgres/blob/d363d42bb9a4399a0207bd3b371c966e22e06bd3/src/include/nodes/parsenodes.h#L248-L269

File: psqlparse/nodes/parsenodes.py [hunk 7]
```diff
@@ -284,3 +313,186 @@ def __init__(self, obj):
...
+        self.type_mod = obj.get('typemod')
+        self.array_bounds = build_from_item(obj, 'arrayBounds')
+        self.location = obj.get('location')
+
+    def tables(self):
+        return set()
+
+
+class SortBy(Node):
+    DIR_DEFAULT = 0
```
Comment (alculquicondor) [inline psqlparse/nodes/parsenodes.py]: constants https://github.com/postgres/postgres/blob/d363d42bb9a4399a0207bd3b371c966e22e06bd3/src/include/nodes/parsenodes.h#L47

File: psqlparse/nodes/parsenodes.py [hunk 8]
```diff
@@ -284,3 +313,186 @@ def __init__(self, obj):
...
+        self.name = Name.from_string(obj.get('name'))
+        self.refname = Name.from_string(obj.get('refname'))
+        self.partition_clause = build_from_item(obj, 'partitionClause')
+        self.order_clause = build_from_item(obj, 'orderClause')
+        self.frame_options = obj.get('frameOptions')
+        self.start_offset = build_from_item(obj, 'startOffset')
+        self.end_offset = build_from_item(obj, 'endOffset')
+        self.location = obj.get('location')
+
+    def tables(self):
```
Comment (alculquicondor) [inline psqlparse/nodes/parsenodes.py]: Are you sure about this override? See https://github.com/alculquicondor/psqlparse/blob/master/psqlparse/nodes/nodes.py#L8 and don't implement the method if you are not sure about the node containing tables (for now). Do the same for any other Node.

File: psqlparse/nodes/primnodes.py [hunk 9]
```diff
@@ -1,9 +1,18 @@
...
 from .nodes import Node
+from .value import Name
+
+
+# See https://github.com/postgres/postgres/blob/master/src/include/nodes/primnodes.h
 
 
 class RangeVar(Node):
 
+    # NB: these changed in PG 10!
```
Comment (alculquicondor) [inline psqlparse/nodes/primnodes.py]: maybe we should add in the readme that we are targeting postgres 9.5

File: psqlparse/nodes/utils.py [hunk 10]
```diff
@@ -11,14 +11,15 @@ def get_node_class(class_name):
     return getattr(module, class_name, None)
 
 
-def build_from_obj(obj):
+def build_from_obj(obj, override_class_name=None):
```
Comment (alculquicondor) [inline psqlparse/nodes/utils.py]: We shouldn't be needing to override the class name if we only have the classes from postgres

File: psqlparse/peace.py [hunk 11]
```diff
@@ -0,0 +1,846 @@
+# -*- coding: utf-8 -*-
```
Comment (alculquicondor) [inline psqlparse/peace.py]: TODO for me: review this code (but ideally it should be a separate PR).

File: tox.ini [hunk 12]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
```
Comment (alculquicondor) [inline tox.ini]: I don't see this being used on the travis configuration, but maybe I'm missing something?

**Comment by lelit** [comment]
Please ignore/close this PR: as requested in #6, I extracted the additional nodes I implemented here and rebased them in a different PR (see #23). That obviously implies the commits presented here will not be valid anymore, and will require further work.
